File: solutions/javascript/yacht/1/yacht.js
const ones = (dice) => dice.reduce((count, die) => count + (die === 1), 0);
const twos = (dice) => 2 * dice.reduce((count, die) => count + (die === 2), 0);
const threes = (dice) => 3 * dice.reduce((count, die) => count + (die === 3), 0);
const fours = (dice) => 4 * dice.reduce((count, die) => count + (die === 4), 0);
const fives = (dice) => 5 * dice.reduce((count, die) => count + (die === 5), 0);
const sixes = (dice) => 6 * dice.reduce((count, die) => count + (die === 6), 0);
const partition = (dice) => {
  const fst = dice.filter((die) => die === dice[0]);
  const snd = dice
    .filter((die) => die !== dice[0])
    .filter((die, _, self) => die === self[0]);
  return [fst, snd];
};
const full = (dice) => {
  const [fst, snd] = partition(dice);
  if (Math.max(fst.length, snd.length) !== 3) return 0;
  return fst.length * fst[0] + snd.length * snd[0];
};
const four = (dice) => {
  const [fst, snd] = partition(dice);
  if (fst.length >= 4) return fst[0] * 4;
  if (snd.length >= 4) return snd[0] * 4;
  return 0;
};
const little = (dice) => 30 * dice.sort().every((die, idx) => die === idx + 1);
const big = (dice) => 30 * dice.sort().every((die, idx) => die === idx + 2);
const choice = (dice) => dice.reduce((acc, die) => acc + die);
const yacht = (dice) => 50 * dice.every((die) => die === dice[0]);

export const score = (dice, category) => {
  const dispatchTable = [
    ones, twos, threes,
    fours, fives, sixes,
    full, four, little,
    big, choice, yacht,
  ];
  return dispatchTable.find((fn) => category.startsWith(fn.name))(dice);
};

File: solutions/cpp/yacht/1/yacht.cpp
#include "yacht.h"

namespace yacht {
    constexpr auto are_equal(const char* first, const char* second) {
        if (*first == '\0' and *second == '\0') return true;
        if (*first == '\0' or *second == '\0' or *first != *second) return false;
        return *first == *second and are_equal(first + 1, second + 1);
    }

    constexpr auto list_of_categories = std::array{
        "ones",
        "twos",
        "threes",
        "fours",
        "fives",
        "sixes",
        "full house",
        "four of a kind",
        "little straight",
        "big straight",
        "choice",
        "yacht"
    };  

    enum class Category {
        ones,
        twos,
        threes,
        fours,
        fives,
        sixes,
        full_house,
        four_of_a_kind,
        little_straight,
        big_straight,
        choice,
        yacht
    };

    constexpr auto num_of_categories = list_of_categories.size();

    constexpr auto check_category(const char* category, std::size_t pos) -> Category {
        return pos == num_of_categories or are_equal(category, list_of_categories[pos]) ? 
                static_cast<Category>(pos) : 
                check_category(category, pos + 1);
    }

    using map_of_frequency = std::array<int, 6>;

    constexpr auto get_dice_frequency(const dice& roll) -> map_of_frequency {
        auto map = map_of_frequency{};
        for (const auto& die : roll)
            map[die - 1]++;
        return map;
    }

    auto score_function(Category category) -> std::function<int(const dice&)> {
        switch (category) {
            case Category::ones:
            case Category::twos:
            case Category::threes:
            case Category::fours:
            case Category::fives:
            case Category::sixes:
                return [multiplier = static_cast<int>(category) + 1](const dice& roll) {
                    auto sum = 0;
                    for (const auto& die : roll)
                        if (die == multiplier) sum += multiplier;
                    return sum;
                };
            case Category::full_house:
                return [] (const dice& roll) {
                    const auto frequencies = get_dice_frequency(roll);
                    auto sum = 0;
                    
                    for (auto index = 0; index < 6; ++index) {
                        const auto number = index + 1;
                        const auto frequency = frequencies[index];
                        if (frequency) {
                            if (frequency != 3 and frequency != 2) return 0;
                            sum += number * frequency;
                        }
                    }                
                    return sum;
                };
            case Category::four_of_a_kind:
                return [] (const dice& roll) {
                    const auto frequencies = get_dice_frequency(roll);
    
                    for (auto index = 0; index < 6; ++index) {
                        const auto number = index + 1;
                        const auto frequency = frequencies[index];
                        if (frequency == 4 or frequency == 5) return number * 4;
                    }
     
                    return 0;
                };
            case Category::little_straight:
                return [] (const dice& roll) {
                    const auto frequencies = get_dice_frequency(roll);
    
                    for (auto index = 0; index < 5; ++index) {
                        const auto frequency = frequencies[index];
                        if (frequency != 1) return 0;
                    }
    
                    return 30;
                };
            case Category::big_straight:
                return [] (const dice& roll) {
                    const auto frequencies = get_dice_frequency(roll);
    
                    for (auto index = 1; index < 6; ++index) {
                        const auto frequency = frequencies[index];
                        if (frequency != 1) return 0;
                    }
    
                    return 30;
                };
            case Category::choice:
                return [] (const dice& roll) {
                    return std::accumulate(
                        roll.begin(),
                        roll.end(),
                        0
                    );
                };
            default:
                return [] (const dice& roll) {
                    const auto frequencies = get_dice_frequency(roll);
                    for (const auto& frequency : frequencies)
                        if (frequency == 5) return 50;
                    return 0;
                };
        }
    }
        
    auto score(const dice& roll, const char* category_string) -> int {
        const auto category = check_category(category_string, 0);
        return score_function(category)(roll);
    }
}  // namespace yacht
